src/mono/wasm/runtime/binding_support.js

```diff
@@ -162,49 +160,174 @@ var BindingSupportLib = {
 			this._js_owned_object_registry = new FinalizationRegistry(this._js_owned_object_finalized.bind(this));
 		},
 
-		_get_weak_delegate_from_handle: function (id) {
-			var result = null;
+		_js_owned_object_finalized: function (gc_handle) {
+			// The JS function associated with this gc_handle has been collected by the JS GC.
+			// As such, it's not possible for this gc_handle to be invoked by JS anymore, so
+			//  we can release the tracking weakref (it's null now, by definition),
+			//  and tell the C# side to stop holding a reference to the managed delegate.
+			this._js_owned_object_table.delete(gc_handle);
+			this.release_js_owned_object_by_handle(gc_handle);
+		},
+
+		_lookup_js_owned_object: function (gc_handle) {
+			if (!gc_handle)
+				return null;
+			var wr = this._js_owned_object_table.get(gc_handle);
+			if (wr) {
+				return wr.deref();
+				// TODO: could this be null before _js_owned_object_finalized was called ?
+				// TODO: are there race condition consequences ?
+			}
+			return null;
+		},
+
+		_wrap_js_thenable_as_task: function (thenable) {
+			this.bindings_lazy_init ();
+			if (!thenable)
+				return null;
+
+			// hold strong JS reference to thenable while in flight
+			// ideally, this should be hold alive by lifespan of the resulting C# Task, but this is good cheap aproximation
+			var thenable_js_handle = this.mono_wasm_free_list.length ? this.mono_wasm_free_list.pop() : this.mono_wasm_ref_counter++;
```

Comment:
I really don't love the intentional ID reuse here, I think we should consider not doing it unless it gets us something in particular. Maybe it does? My guess is it's because the object registry is an array instead of a Map so they want to keep it dense, but I'm not sure that actually justifies itself since you could end up with a very big free list and a registry full of holes.

Reply:
Agreed, let's do that in separate PR. Also we could consider switching to `Map`

Reply:
https://github.com/dotnet/runtime/issues/56965